java-client/barrage/src/main/java/io/deephaven/client/impl/BarrageSubscriptionImpl.java

```diff
@@ -203,130 +225,51 @@ public BarrageTable partialTable(RowSet viewport, BitSet columns, boolean revers
-                completedCondition.await();
-            }
-        } else {
-            synchronized (this) {
-                while (!checkIfCompleteOrThrow()) {
-                    wait(); // BarrageSubscriptionImpl lock
-                }
-            }
+    private boolean onFutureCancel(boolean cancelResult) {
+        if (!cancelResult) {
+            return false;
         }
 
-        return resultTable;
+        cancel();
+        return true;
     }
 
-    private synchronized void signalCompletion() {
-        completed = true;
-
+    private void onFutureComplete() {
         // if we are building a snapshot via a growing viewport subscription, then cancel our subscription
-        if (isSnapshot) {
-            observer.onCompleted();
+        if (isSnapshot && tryRecordDisconnect()) {
+            GrpcUtil.safelyCancel(observer, "Barrage snapshot is complete", null);
         }
-
-        if (completedCondition != null) {
-            resultTable.getUpdateGraph().requestSignal(completedCondition);
-        }
-        notifyAll();
-    }
-
-    @Override
-    public BarrageTable snapshotEntireTable() throws InterruptedException {
-        return snapshotEntireTable(true);
-    }
-
-    @Override
-    public BarrageTable snapshotEntireTable(boolean blockUntilComplete) throws InterruptedException {
-        return snapshotPartialTable(null, null, false, blockUntilComplete);
-    }
-
-    @Override
-    public BarrageTable snapshotPartialTable(RowSet viewport, BitSet columns) throws InterruptedException {
-        return snapshotPartialTable(viewport, columns, false, true);
-    }
-
-    @Override
-    public BarrageTable snapshotPartialTable(RowSet viewport, BitSet columns, boolean reverseViewport)
-            throws InterruptedException {
-        return snapshotPartialTable(viewport, columns, reverseViewport, true);
-    }
-
-    @Override
-    public BarrageTable snapshotPartialTable(RowSet viewport, BitSet columns, boolean reverseViewport,
-            boolean blockUntilComplete) throws InterruptedException {
-        isSnapshot = true;
-        return partialTable(viewport, columns, reverseViewport, blockUntilComplete);
     }
 
     @Override
     protected void destroy() {
         super.destroy();
-        close();
+        cancel();
     }
 
-    private void handleDisconnect() {
-        if (!connected) {
+    private void cancel() {
+        if (!tryRecordDisconnect()) {
             return;
         }
-        // log an error only when doing a true subscription (not snapshot)
-        if (!isSnapshot) {
-            log.error().append(this).append(": unexpectedly closed by other host").endl();
-        }
-        cleanup();
-    }
 
-    @Override
-    public synchronized void close() {
-        if (!connected) {
-            return;
-        }
-        GrpcUtil.safelyComplete(observer);
+        resultTable.handleBarrageError(new RequestCancelledException("Barrage subscription cancelled by client"));
```

Comment:
We don't need this handleBarrageError on a future cancel, right? Nothing can get the Table in that case. And we don't really need it on a destroy, because the table had better also be dead at that point; we just need to complete the future exceptionally, if it's not already complete.

Reply:
I think this one we want if `!isSnapshot`. Maybe we want to `resultTable.forceReferenceCountToZero()` instead?